Function extension mechanism and several functions (#280)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

The logical AND, OR, and NOT operators have the normal semantics of Boolean algebra and
obey its laws (see, for example, {{BOOLEAN-LAWS}}).

##### Function Extensions {#fnex}
{: unnumbered}

[^unnumbered-bad]

[^unnumbered-bad]: This should not be an unnumbered section, which it is
    forced to be by the parent section being unnumbered too.
    The overall structure where the parts of this subsection go to
    needs to be decided; one part already is in the IANA
    Considerations in {{iana-fnex}}
{:source=" -- cabo"}

Beyond the filter expression functionality defined in the preceding
subsections, JSONPath defines an extension point that can be used to
add filter expression functionality: "Function Extensions".

A function extension defines a registered name (see {{iana-fnex}}) that
can be applied to a sequence of zero or more arguments, resulting in a result.
A function-expression stands for ("returns") a nodelist or a value.
The arguments are filter expressions that are to be interpreted as
nodelists or values.
For each argument and for the result, the function extension defines
the "kind" of the item, i.e., whether the item is a value or a

> nodelist ("nodes").

Reviewer: Perhaps rather than the item being  a value, it could be a singular nodelist. This allows chaining the result of a function as an argument of another function with propagation of the empty nodelist. Whether we make a general rule that empty argument nodelists MUST produce empty result nodelists or whether this is a per-function optional behaviour is something we should decide. I quite like the uniformity and predictability of a general rule.

Reviewer 2: Shouldn't the function definition being able to define the "return" even more concretely as a string or number? The latter would be, what is expected with our `length` function.

Reviewer: Also, what happens when a function is passed an argument of the wrong kind? Does this produce a syntax error? Is there any coercion to the correct kind? Are there any "soft" runtime errors in which case what does the function return?

Author: Let's see which mistakes can be caught at compile time, and whether we can stick with "no runtime errors".

~~~ abnf
function-name           = function-name-first *function-name-char
function-name-first     = LCALPHA

> function-name-char      = DIGIT / name-first / "_"

Reviewer: s /name-first/function-name-first/

> function-name-char      = DIGIT / name-first / "_"

Reviewer: ```suggestion
function-name-char      = DIGIT / function-name-first / "_"
```

LCALPHA                 = %x61-7A  ; "a".."z"

> function-expression     = function-name "(" [*(function-argument "," S) function-argument] ")"

Reviewer: Do we want to allow more whitespace, e.g. just inside the parentheses?

Author: space inside parentheses is evil.  Except for line breaks...

Author: Well, the answer is yes; we should move any attempt to reduce variability in spacing to its own PR.

> function-expression     = function-name "(" [*(function-argument "," S) function-argument] ")"

Reviewer: The following is more consistent with the rest of our ABNF:
```suggestion
function-expression     = function-name "(" [function-argument *("," S function-argument)] ")"
```

> singular-path           =/ function-expression

Reviewer: Aside: I like incremental alternatives and if we keep them here, we may want to use them elsewhere, e.g. in the definition of `selector`.

Author: This was an expedient device to fit this in without touching everything -- should be fixed IMHO

> singular-path           =/ function-expression

Reviewer: If a function returns a nodelist, how can the result be used as a singular path?

Author: That is certainly a part that we still need to address.
Functions that return a nodelist probably need to be categorized into singular ones and ones that are not.
Let's to that when we agree on the overall architecture of the extension.

function-argument       = comparable
~~~

Syntactically, a function-expression can occur anywhere where a
singular-path can occur (exist-expr, comparable, regex-expr).
A function-expression that employs a function extension that returns a

> nodelist MUST return a singular node to be used in a comparable or a

Reviewer: Rather than a singular node, perhaps return a singular nodelist (see below).

> regex-expr.

Reviewer: If we fix #195, I think the language here can be simplified. Until then, we need to include `exist-expr` in the places where a singular node must be returned.

Reviewer 2: I don't think, that #195 is relevant here, since `exist-expr` always results in a Boolean, regardless of the presence of a singular-path or not.

Reviewer: My point was about an `exist-expr` which tests the existence of the results of a function. Unless/until #195 is merged, a function used in such a context MUST return a singular node.

Author: Let's revisit this with the complete proposal

Author: grammar fixed in 08dd7ad
now add semantic restriction

###### Function Extensions: length {#length}
{: unnumbered}

When defining an extension point in a protocol, it is generally a good
idea to provide at least one extension that makes use of the extension
point, to serve as an example and to provide motivation to implement

> the extension point.

Reviewer 3: I don't know that this is need in the spec.  Might be good for a comment, but its target audience is really people who write specs.

Author: (Maybe I should have added a "to me removed by RFC editor" comment.)

Author: Removed in f7295bd

The "length" function extension provides a way to compute the length
of an array and make that available for further processing in the
filter expression:

~~~ JSONPath
$[?length(@.authors) >= 5]
~~~

Its only argument is a value (possibly taken from a singular path as
in the example above).  The result also is a value; if the argument

> value is an array, the result is the number of elements in the array.

Reviewer 3: Do we also want to support strings or objects?  They're fairly well-defined for this function.

Author: Very good question.  What do implementations out there do?
For strings, I think we'd count the Unicode scalars.
For JSON objects, the number of name-value pairs.

Reviewer 2: We even could allow nodelists and return the number of nodes.

Author: That would need to be a separate extension, as length(X) is ambiguous if X can be interpreted as a node that contains an array (which gets us the size of the array) or a nodelist of size 1 (which gets us 1).

Author: Separate function (called "count") provided now.

[length-mismatch]

> [length-mismatch]: Define the case where the argument is not an array as well.

Reviewer: It seems we have limited options if we are going to be consistent with the rest of the spec. We could either return a fixed value (e.g. `-1`) or an empty nodelist. The latter seems preferable as it's consistent with a singular path possibly being empty.

This makes me wonder whether we should make a function always return a nodelist and have the function spec say whether or not the returned nodelist is "singular" (meaning it has at most one member) instead of allowing nodelists of values to be returned.

Author: The main problem is that the node that has the length isn't in the input JSON value

Reviewer: Of course. I was thinking "list of nodes" rather than _nodelist_.

Reviewer: If we can avoid the function having to create nodes, so much the better. So I guess that limits return values to nodelists or values after all.

Reviewer 2: I like the well known return value `-1` for the case of an invalid argument.

Author: I made a different proposal specifically for length(), but we do need to come up with a more general solution.

                             *(S "," S function-argument)] S ")"
singular-path           =/ function-expression
function-argument       = comparable
~~~

Syntactically, a function-expression can occur anywhere where a
singular-path can occur (exist-expr, comparable).
A function-expression that employs a function extension that returns a
nodelist MUST return a singular node to be used in a comparable.

###### Function Extensions: length {#length}
{: unnumbered}

Arguments:
: 1. value

Result:
: value (unsigned integer)

The "length" function extension provides a way to compute the length
of a value and make that available for further processing in the
filter expression:

~~~ JSONPath
$[?length(@.authors) >= 5]
~~~

Its only argument is a value (possibly taken from a singular path as
in the example above).  The result also is a value, an unsigned
integer.

* If the argument value is a string, the result is the number of
  Unicode scalar values in the string.
* If the argument value is an array, the result is the number of
  elements in the array.
* If the argument value is an object, the result is the number of
  members in the object.

> * For any other argument value, the result is one.

Reviewer: Does that include the case of a singular path which returns an empty nodelist?

Author: That is a very good question.  As defined, the argument value needs to be a value, so handling empty nodelists is not defined.  But it should be possible to do this...

> ###### Function Extensions: count {#count}

Reviewer 3: What's the difference between `count()` and `length()`?  Why have a distinction?

Author: ``` 
###### Function Extensions: length {#length}
{: unnumbered}

Arguments:
: 1. value

###### Function Extensions: count {#count}
{: unnumbered}

Arguments:
: 1. nodelist
```

One is for JSON values, one is for nodelists.

Reviewer 3: This isn't apparent to me from the text.  Some additional clarification would be good. (May be provided with the examples I asked for in another comment.)

{: unnumbered}

Arguments:
: 1. nodelist

Result:
: value (unsigned integer)

The "count" function extension provides a way to obtain the number of
nodes in a nodelist and make that available for further processing in
the filter expression:

~~~ JSONPath

> $[?count(@.*.author) >= 5]

Reviewer: This seems syntactically invalid since any function argument has to be a `comparable` which `@.*.author` isn't.

Author: Yes.  This doesn't have the results of merging #195 incorporated.

~~~

Its only argument is a nodelist.
The result is a value, an unsigned integer, that gives the number of
nodes in the nodelist.
Note that there is no deduplication of the nodelist. [^dedup]

> [^dedup]: Well, that can be discussed.

Reviewer: Perhaps add a remdup/dedup function?

Author: Maybe.  Maybe that is a bit too much invention for the base standard.  I myself don't have an opinion on this.

> ###### Function Extensions: match {#match}

Reviewer 3: Since we have `match()` and `search()` now, we should probably remove `=~`.

Reviewer: It has mostly gone. There are just two vestigial mentions left.

Author: Fixed in 52ca26b

: 1. value (string)
  2. value (string conforming to {{-iregexp}})

Result:
: value (`true` or `false`)

The "match" function extension provides a way to check whether a given
string matches a given regular expression, which is in {{-iregexp}} form.

~~~ JSONPath
$[?match(@.date, "1974-05-..")]
~~~

Its first argument is a string that is matched against the iregexp
contained in the string that is the second argument.
The result is `true` if the string matches the iregexp and `false`
otherwise.

###### Function Extensions: search {#search}
{: unnumbered}

Arguments:
: 1. value (string)
  2. value (string conforming to {{-iregexp}})

Result:
: value (`true` or `false`)

The "search" function extension provides a way to check whether a
given string contains a substring that matches a given regular
expression, which is in {{-iregexp}} form.

~~~ JSONPath
$[?search(@.author, "[BR]ob")]
~~~

Its first argument is a string that is searched for at least one
substring that matches against the iregexp contained in the string
that is the second argument.
The result is `true` if such a substring exists, `false` otherwise.

o o o o o o o o o o o o o o o o o o o o o o o o o o o o o o o o

> #### Examples

Reviewer 3: Need examples of `match()` and `search()` to better highlight the difference between them.  The text varies only by one saying "contains a match" vs just "matches," which I don't think is clear.  In my experience with regexes, containing a match _is_ matching.

Author: Right.  The one example that has a match() would work the same with a search()... we probably need to set up a pair of contrasting examples.  Let's do a round through the examples once the technical content has stabilized.

Author: The text is fine though.  87d95d6 adds a few words that might help grasping the difference.

A function extension defines a registered name (see {{iana-fnex}}) that
can be applied to a sequence of zero or more arguments, resulting in a
result.
A function-expression stands for ("returns") a nodelist or a value.
The arguments are filter expressions that are to be interpreted as
nodelists or values.
For each argument and for the result, the function extension defines
the "kind" of the item, i.e., whether the item is a value or a
nodelist ("nodes").

~~~ abnf
function-name           = function-name-first *function-name-char
function-name-first     = LCALPHA
function-name-char      = DIGIT / function-name-first / "_"
LCALPHA                 = %x61-7A  ; "a".."z"

function-expression     = function-name "(" S [function-argument
                             *(S "," S function-argument)] S ")"
singular-path           =/ function-expression ; if value
filter-path             =/ function-expression ; if nodelist
function-argument       = comparable / filter-path
~~~

Syntactically, a function-expression can occur anywhere where a
singular-path can occur (exist-expr, comparable).
A function-expression that employs a function extension that returns a

> nodelist MUST return a singular node to be used in a comparable.

Reviewer: Why would it not be acceptable to return an empty nodelist since that's valid for use in a comparable?

Reviewer: Separately: what should the behaviour be if a nodelist with multiple nodes is returned in this situation?

Author: One way to handle this would be to build a type system with values, singular paths (which are implicitly converted into values where a value is required), and filter-paths.
Functions already have to define the types of their arguments and their result, so this could be checked statically.

Reviewer: Yes, I like that. The more static checks we can get, the fewer runtime errors the spec will need to cope with.

| `$[?@[?@.b]]` | `[3, 5, 1, 2, 4, 6, {"b": "j"}, {"b": "k"}, {"b": {}}]` | `$['a']` | Nested filters |
| `$.a[?@.b, ?@.b]` | `{"b": "j"}` <br> `{"b": "k"}` <br> `{"b": "k"}` <br> `{"b": "j"}` | `$['a'][6]` <br> `$['a'][7]` <br> `$['a'][7]` <br> `$['a'][6]` | Non-deterministic ordering |
| `$.a[?@<2 || @.b == "k"]` | `1` <br> `{"b": "k"}` | `$['a'][2]` <br> `$['a'][7]` | Array value logical OR |

> | `$.a[?match(@.b, "[jk]")]` | `{"b": "j"}` <br> `{"b": "k"}` | `$['a'][6]` <br> `$['a'][7]` | Array value regular expression |

Reviewer 3: Can we get one with `search()` as well, please?

Provisional registration? (standards tree only):
: no

## Function Extensions {#iana-fnex}

This specification defines a new "Function Extensions sub-registry" in
a new "JSONPath Parameters registry", with the policy "expert review"
({{Section 4.5 of -ianacons}}).

> The expert is instructed to be frugal in the allocation of very short

Reviewer: Is there just one expert? Why not:
```suggestion
Experts are instructed to be frugal in the allocation of very short
```

Author: ```suggestion
The experts is instructed to be frugal in the allocation of very short
```

I'd say.

> function extension names, keeping them in reverse for applications that

Reviewer: ```suggestion
function extension names, keeping them in reserve for applications that
```

> function extension names, keeping them in reserve for applications that

Reviewer 3: What do you mean by "applications" here?

Author: application in the sense of the way this is going to the used, usage

Reviewer 3: I wonder if another word would work better here so that readers don't conflate it with executable applications or other consuming code.

Reviewer: Perhaps:
```suggestion
function extension names, keeping them in reserve for functions that
```

> are likely to enjoy wide use and can make good use of their shortness.

Reviewer: Not sure about "can make good use of their shortness". When would a short name not be made good use of?

Author: Indeed, this is really less about shortness.  Say, someone registers a size() that extracts the shirt size of a JSON description of a shirt.

Reviewer: Actually, why not simply the following?
```suggestion
are likely to enjoy wide use.
```
You see I'm not sure the phrase "and can make good use of their shortness" is anything more than motherhood.

Author: Addressed in 85e5525

The expert is also instructed to direct the registrant to provide a
specification ({{Section 4.6 of -ianacons}}), but can make exceptions,
for instance when a specification is not available at the time of

> registration but is likely forthcoming.

Reviewer: This may be common policy for IANA, but the risk is that a spec may never appear and we end up with people "squatting" on function names so others can't register them.

Reviewer 3: Perhaps the function name reservation can have an expiry.

Author: Finally, I can reply to this observation.
Copying from below:

Expiry is somewhat unusual for IANA registries (early allocations do expire).
What is the behavior you would expect implementations to exhibit?
The function extension to stop working on the expiry day?

Reviewer: @gregsdennis I think those questions are for you. (I wasn't suggesting expiry.)

Author: I think the DE instructions are open enough so that the DE can be quite insisting on a specification, so I don't see such a potential for disaster here.

If the expert becomes aware of function extensions that are deployed and
in use, they may also initiate a registration on their own if
they deem such a registration can avert potential future collisions.
{: #de-instructions}

Each entry in the registry must include:

{:vspace}
Function Name:
: a lower case ASCII {{-ascii}} string that starts with a letter and can
  contain letters, digits and underscore characters afterwards

>   (`[a-z][_a-z0-9]*`).

Reviewer 3: So we're only accepting English-character function names?  Just clarifying.

Author: Well, ASCII, yes.
This is an interesting question in a more general way, as so far JSONPath is mostly indifferent to languages.
But these function extensions are intended for an international audience, so it is probably best to do a lowest common denominator.

Author: Obviously, I stole the ABNF from dot-member-name, which allows %x80-10FFFF -- this is a recipe for confusion, with RTL characters, format effectors, and all that.  I'm having a hard time making a suggestion as to what we should allow there -- `\p{L}|\p{N}` maybe, in parallel to allowing ALPHA + DIGIT?  But then we also allow "_", and there may be equivalent separators in some scripts.  So this is hard to get useful over a wide domain, while reducing the potential for confusion.  function names of course do not depend on whatever language was used in the input data item, so we can be more frugal in what we allow.

Author: No further discussion -- I'm assuming we stick with `[a-z][_a-z0-9]*` (no change required)

Brief description:
: a brief description

Input:
: A comma-separated list of zero or more kinds (value or nodes) of the
  arguments expected for this function extension

Output:
: The kind (value or nodes) of the result for this function extension

Change Controller:
: (see {{Section 2.3 of -ianacons}})

Reference:
: a reference document that provides a description of the function
  extension

Initial entries in this sub-registry are as listed in {{pre-reg}}:

| Function Name | Brief description | Input | Output | Change Controller | Reference |   |   |

> | length        | length of array   | value | value  | IESG              | {{fnex}} of RFCthis |   |   |

Reviewer: Do we need to allow an empty nodelist as input so that we can apply `length` to a `singular-path`?

Author: That should be a separate function (for a singular node list, it is not clear whether you want 1 or the length of the value in the node).

Author: ABNF says:

```
function-argument       = comparable
```

So literals are the other case.

Author: Separate function now in my proposal update.
